contracts/oracle/src/contract/alarms.rs

```diff
@@ -1,62 +1,385 @@
+        }
+
+        impl<'m, BaseC> WithQuote<BaseC> for AddAlarms<'m>
+        where
+            BaseC: Currency,
+        {
+            type Output = Response;
+            type Error = ContractError;
+
+            fn exec<C>(self, below: Price<C, BaseC>) -> Result<Self::Output, Self::Error>
+            where
+                C: Currency,
+            {
+                if let Some(above) = self.above {
+                    self.price_alarms.add_alarm_above::<C, BaseC>(
+                        self.storage,
+                        &self.addr,
+                        above.try_into()?,
+                    )?;
+                }
+                self.price_alarms
+                    .add_alarm_below(self.storage, &self.addr, below)?;
+                Ok(Response::new())
+            }
+        }
+
+        let (below, above) = alarm.into();
+        with_quote::execute::<_, _, _, BaseC>(
+            &below,
+            AddAlarms {
+                storage,
+                addr,
+                above,
+                price_alarms: Self::PRICE_ALARMS,
+            },
+        )
     }
 
-    pub fn try_notify_alarms(
+    pub fn try_notify_alarms<BaseC>(
         storage: &mut dyn Storage,
         mut batch: Batch,
         prices: &[SpotPrice],
-        max_count: u32,
-    ) -> Result<Response, ContractError> {
-        let sent = Self::PRICE_ALARMS.notify(storage, &mut batch, prices, max_count)?;
-        Ok(Response::from(batch).set_data(to_binary(&DispatchAlarmsResponse(sent))?))
+        max_count: u32, // TODO: type alias
+    ) -> Result<Response, ContractError>
+    where
+        BaseC: Currency,
+    {
+        let mut next_id = Self::MSG_ID.may_load(storage)?.unwrap_or_default();
+        let initial_id = next_id;
+
+        Self::alarms_iter::<BaseC>(storage, prices)
+            .take(max_count.try_into()?)
+            .try_for_each(|addr| Self::schedule_alarm(&mut batch, addr?, &mut next_id))?;
+
+        Self::MSG_ID.save(storage, &next_id)?;
+
+        let processed = next_id.wrapping_sub(initial_id);
+
+        Ok(Response::from(batch)
+            .set_data(to_binary(&DispatchAlarmsResponse(processed.try_into()?))?))
     }
 
-    pub fn try_query_alarms(
+    pub fn try_query_alarms<BaseC>(
         storage: &dyn Storage,
         prices: &[SpotPrice],
-    ) -> Result<AlarmsStatusResponse, ContractError> {
-        let remaining_alarms = Self::PRICE_ALARMS.query_alarms(storage, prices)?;
-        Ok(AlarmsStatusResponse { remaining_alarms })
+    ) -> Result<AlarmsStatusResponse, ContractError>
+    where
+        BaseC: Currency,
+    {
+        Ok(AlarmsStatusResponse {
+            remaining_alarms: Self::alarms_iter::<BaseC>(storage, prices)
+                .next()
+                .transpose()?
+                .is_some(),
+        })
+    }
+
+    fn alarms_iter<'a, BaseC>(
+        storage: &'a dyn Storage,
+        prices: &'a [SpotPrice],
+    ) -> impl Iterator<Item = Result<Addr, StdError>> + 'a
```

Comment:
StdError -> ContractError?

Reply:
more specifically, the PriceAlarms::alarms should return items that are Result with AlarmError ... to make the marketprice API consistent and to have it use its own error type instead of the one of the internal implementation